src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+	CRITICAL_SECTION critical;  /* probably unneeded, but good practice */
+	STARTUPINFO startup;
+	PROCESS_INFORMATION process;
+	HANDLE hpipe[6] = { NULL, NULL, NULL, NULL, NULL, NULL };
+	int *fd[3] = { stdin_fd, stdout_fd, stderr_fd };
+	const char *failed;     /* failed WIN32/CRTL function, if any */
+	gchar *message = NULL;  /* glib WIN32/CTRL error message */
+	gchar *failure = NULL;  /* full error text */
+	int i;
+
+	ZeroMemory(&startup, sizeof startup);
+	startup.cb = sizeof(startup);
+	startup.dwFlags = STARTF_USESHOWWINDOW;
+	startup.wShowWindow = SW_SHOWMINNOACTIVE;  /* so you can switch to it if needed */
+
+	if (stdin_fd || stdout_fd || stderr_fd)
+	{
+		/* Not all programs accept mixed NULL and non-NULL hStd*, so we create all */
+		startup.dwFlags |= STARTF_USESTDHANDLES;
+
+		for (i = 0; i < 3; i++)
+		{
+			static int pindex[3][2] = { { READ_STDIN, WRITE_STDIN },
+					{ READ_STDOUT, WRITE_STDOUT }, { READ_STDERR, WRITE_STDERR } };
+			DWORD state;
+
+			if (!CreatePipe(&hpipe[pindex[i][0]], &hpipe[pindex[i][1]], NULL, 0))
+			{
+				hpipe[pindex[i][0]] = hpipe[pindex[i][1]] = NULL;
+				failed = "CreatePipe";
+				goto leave;
+			}
+
+			if (fd[i])
+			{
+				static int mode[3] = { _O_WRONLY, _O_RDONLY, _O_RDONLY };
+
+				if (!GetNamedPipeHandleState(hpipe[i], &state, NULL, NULL, NULL, NULL, 0))
+				{
+					failed = "GetNamedPipeHandleState";
+					goto leave;
+				}
+
+				state |= PIPE_NOWAIT;
+				if (!SetNamedPipeHandleState(hpipe[i], &state, NULL, NULL))
+				{
+					failed = "SetNamedPipeHandleState";
+					goto leave;
+				}
+
+				if ((*fd[i] = _open_osfhandle((intptr_t) hpipe[i], mode[i])) == -1)
+				{
+					failed = "_open_osfhandle";
+					message = g_strdup(g_strerror(errno));
+					goto leave;
+				}
+			}
+			else if (!CloseHandle(hpipe[i]))
+			{
+				failed = "CloseHandle";
+				goto leave;
+			}
+
+			if (!SetHandleInformation(hpipe[i + 3], HANDLE_FLAG_INHERIT,
+				HANDLE_FLAG_INHERIT))
+			{
+				failed = "SetHandleInformation";
+				goto leave;
+			}
+		}
+	}
+
+	startup.hStdInput = hpipe[READ_STDIN];
+	startup.hStdOutput = hpipe[WRITE_STDOUT];
+	startup.hStdError = hpipe[WRITE_STDERR];
+
+	if (!CreateProcess(NULL, command_line, NULL, NULL, TRUE, NORMAL_PRIORITY_CLASS,
+		environment, working_directory, &startup, &process))
+	{
+		failed = "CreateProcess";
+		/* no errors beyond this point */
+	}
+	else
+	{
+		failed = NULL;
+		*hprocess = process.hProcess;
+	}
+
+leave:
+	if (failed)
+	{
+		if (!message)
+			message = g_win32_error_message(GetLastError());
+
+		failure = g_strdup_printf("%s() failed: %s", failed, message);
+		g_free(message);
+	}
+
+	if (startup.dwFlags & STARTF_USESTDHANDLES)
+	{
+		for (i = 0; i < 3; i++)
+		{
+			if (failed)
+			{
+				if (fd[i] && *fd[i] != -1)
+					_close(*fd[i]);
+				else if (hpipe[i])
+					CloseHandle(hpipe[i]);
+			}
+
+			if (hpipe[i + 3])
+				CloseHandle(hpipe[i + 3]);
+		}
+	}
+
+	LeaveCriticalSection(&critical);

```

Comment:
There should be. Judging from the comment after the declaration of "critical", I was probably in doubt wheher to use it, and ended with 1/2. :)